examples/threshold_sig.rs

```diff
@@ -0,0 +1,203 @@
+    // `threshold` - our protocol requires a message to have `threshold + 1` validator signatures
+    // before it can be added to the `chat_log`.
+    fn new(n_nodes: usize, threshold: usize) -> Self {
+        let mut rng = rand::thread_rng();
+        let sk_set = SecretKeySet::random(threshold, &mut rng).unwrap();
+        let pk_set = sk_set.public_keys();
+
+        let nodes = (0..n_nodes)
+            .map(|id| {
+                let sk_share = sk_set.secret_key_share(id).unwrap();
+                Node::new(id, sk_share)
+            })
+            .collect();
+
+        ChatNetwork {
+            pk_set,
+            nodes,
+            chat_log: vec![],
+            n_users: 0,
+        }
+    }
+
+    fn create_user(&mut self) -> User {
+        let user_id = self.n_users;
+        let user = User::new(user_id);
+        self.n_users += 1;
+        user
+    }
+
+    fn get_node(&mut self, id: NodeId) -> &mut Node {
+        self.nodes
+            .get_mut(id)
+            .expect("No `Node` exists with that ID")
+    }
+
+    // Run a single round of the consensus algorithm. If consensus produced a new block, append
+    // that block the chat log. 
+    fn step(&mut self) {
+        if let Some(block) = self.run_consensus() {
+            self.chat_log.push(block);
+        }
+    }
+
+    // Our chat protocol's consensus algorithm. Produces a new block to be appended to the chat
+    // log. Our consensus uses threshold-signing to verify that a message has received enough
+    // signature shares (i.e. has been signed by `threshold + 1` nodes).
+    fn run_consensus(&self) -> Option<(UserId, Msg, Signature)> {
+        // Create a new `MsgDatabase` of every message that has been signed by a validator node.
+        let all_pending: MsgDatabase = self.nodes.iter()
+            .fold(BTreeMap::new(), |mut all_pending, node| {
+                for (user_id, signed_msgs) in node.pending.iter() {
+                    let mut user_msgs = all_pending
+                        .entry(*user_id)
+                        .or_insert(BTreeMap::new());
+                    for (msg, sigs) in signed_msgs.iter() {
+                        let sigs: Vec<NodeSignature> = sigs.iter().map(|sig| sig.clone()).collect();
+                        user_msgs
+                            .entry(msg.to_string())
+                            .or_insert(Vec::new())
+                            .extend(sigs);
+                    }
+                }
+                all_pending
+            });
+
+        // Iterate over the `MsgDatabase` numerically by user ID, then iterator over each user's
+        // messages alphabetically. Try to combine the validator signatures. The first message that
+        // has received `threshold + 1` node signatures, will produce a valid "combined" signature
+        // and will be added to the chat log.
+        for (user_id, signed_msgs) in all_pending.iter() {
```

Comment:
Maybe `in &all_pending` is more idiomatic (also below). Clippy prefers it, I think.